nbclient/client.py

```diff
@@ -607,22 +642,33 @@ def _get_timeout(self, cell):
 
         return timeout
 
-    async def _async_handle_timeout(self, timeout, cell=None):
+    async def _async_handle_timeout(
+            self,
+            timeout: int,
+            cell: t.Optional[NotebookNode] = None) -> None:
+
         self.log.error("Timeout waiting for execute reply (%is)." % timeout)
         if self.interrupt_on_timeout:
             self.log.error("Interrupting kernel")
+            assert self.km is not None
             await ensure_async(self.km.interrupt_kernel())
         else:
             raise CellTimeoutError.error_from_timeout_and_cell(
                 "Cell execution timed out", timeout, cell
             )
 
-    async def _async_check_alive(self):
+    async def _async_check_alive(self) -> None:
+        assert self.kc is not None
         if not await ensure_async(self.kc.is_alive()):
             self.log.error("Kernel died while waiting for execute reply.")
             raise DeadKernelError("Kernel died")
 
-    async def async_wait_for_reply(self, msg_id, cell=None):
+    async def async_wait_for_reply(
+            self,
+            msg_id: str,
+            cell: t.Optional[NotebookNode] = None) -> t.Optional[t.Dict]:
+
+        assert self.kc is not None
```

Comment:
A little odd to keep the asserts in non-test code.

Reply:
Here too, with the assert we get `error: Item "None" of "Optional[Any]" has no attribute "shell_channel"` later in the method, because the type checker has no idea `self.kc` has a non-`None` value at that time.